ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,410 @@
+        pure $ case (actual, expected) of
+            (ReadThisJudgment x, ReadThisJudgment y) ->
+                QC.counterexample "lastCheck" $ x QC.=== y
+            _ ->
+                error "impossible! lastCheck response"
+
+    watcherEvents <- dumpEvents varEvents
+
+    pure
+      $ QC.monadicIO
+      $ QSM.prettyCommands prettySM hist
+      $ QC.counterexample
+            (unlines
+               $ (:) "WATCHER"
+               $ map (("    " <>) . show)
+               $ watcherEvents
+            )
+      $ QC.tabulate
+            "Notables"
+            (case Set.toList $ mNotables model' of
+               []       -> ["<none>"]
+               notables -> map show notables
+            )
+      $ QSM.checkCommandNames cmds
+      $ noExn QC..&&. lastCheck QC..&&. res QC.=== QSM.Ok
+
+-----
+
+durationUntilTooOld :: Selection -> IOSim.IOSim s GSM.DurationFromNow
+durationUntilTooOld sel = do
+    let expiryAge = ageLimit `SI.addTime` onset sel
+    now <- SI.getMonotonicTime
+    pure $ case compare expiryAge now of
+        LT -> GSM.Already
+        GT -> GSM.After $ realToFrac $ expiryAge `SI.diffTime` now
+
+        -- 'boringDur' cannot prevent this case. In particular, this case
+        -- necessarily arises in the GSM itself during a 'TimePasses' that
+        -- incurs a so-called /flicker/ event, in which the anti-thrashing
+        -- timer expires and yet the node state at that moment still
+        -- _immediately_ indicates that it's CaughtUp. For the specific case of
+        -- this test suite, the answer here must be 'GSM.Already'.
+        EQ -> GSM.Already
+
+-----
+
+-- | Ensure the GSM thread's transactions quiesce
+--
+-- I'm unsure how many 'yield's are actually necessary, but ten is both small
+-- and also seems likely to suffice.
+--
+-- Despite the crudeness, this seems much more compositional than invasive
+-- explicit synchronization.
+yieldSeveralTimes :: MonadFork m => m ()
+yieldSeveralTimes = replicateM_ 10 yield
+
+{-
+
+Note [Why yield after the command]
+
+For this 'prop_sequential1' repro
+
+@
+   YoungEnough
+
+   Command (NewCandidate Amara (B 1)) Unit []
+   Command (StartIdling Amara) Unit []
+   Command (TimePasses 61) Unit []
+   Command (ExtendSelection (S (-4))) Unit []
+   Command ReadMarker (ReadThisMarker Absent) []
+
+   (Command ReadJudgment _ [])   -- this last command is baked into the property
+@
+
+If we yield after the command, then both GSM flicker writes happen during the
+'ExtendSelection'.
+
+If we yield before the command, then both GSM flicker writes happen during the
+'ReadMarker'.
+
+If we don't yield, one write happens during the ReadMarker and the other
+happens /between/ 'ReadMarker' and 'ReadJudgment'.
+
+It seems most intuitive for the updates to happen "as part of" the
+'ExtendSelection', so I'm favoring yielding after.
+
+And since we're yielding after the command, we should also yield before the
+first command, for consistency.
+
+-}
```

Comment:
I tried to convert this to a regression test:
```haskell
prop_yield :: QC.Property
prop_yield =
   prop_sequential1 YoungEnough $ QSM.Commands
     [ QSM.Command (NewCandidate Amara (B 1)) Unit []
     , QSM.Command (StartIdling Amara) Unit []
     , QSM.Command (TimePasses 61) Unit []
     , QSM.Command (ExtendSelection (S (-4))) Unit []
     , QSM.Command ReadMarker (ReadThisMarker Absent) []
     ]
```
First, I had to apply this diff to get a useful counterexample output:
```diff
diff --git a/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs b/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs
index 4d34fb8ed..066b2171c 100644
--- a/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs
+++ b/ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs
@@ -195,7 +195,7 @@ prop_sequential1 j0 cmds = runSimQC $ do
     let executionSM = sm (Just maxBound) vars j0
 
         -- NB the specific IO type is unused here
-        prettySM = sm undefined undefined undefined
+        prettySM = sm undefined undefined j0
 
     let gsm = GSM.realGsmEntryPoints GSM.GsmView {
             GSM.antiThunderingHerd = Nothing
```
After that, the test failed due to a precondition violation (non-boring selection). However, that precondition is actually semantically irrelevant (https://github.com/IntersectMBO/ouroboros-consensus/pull/808/#discussion_r1487577686), so I disabled it.

Then, the tests pass both with the current definition of `yieldSeveralTimes` as well as when replacing it with `pure ()`.

Initially, I thought this note was trying to justify why `yieldSeveralTimes` is necessary to get the tests to pass. However, I now think it is "just" used to get nicer counterexample output. Can you confirm and maybe mention this explicitly?

Reply:
> Can you confirm and maybe mention this explicitly?

I believe the `yield` should be necessary. It seems so, for me --- which should be true for everyone, since `IOSim` is deterministic.

> I tried to convert this to a regression test:

Great point! I've done so now.

> ... non-boring selection ...

We chatted on the call about this; there was a misunderstanding. _Morever_, though, `boringDur` was too conservative --- I've fixed that, and now this repro is accepted.

> Then, the tests pass both with the current definition of yieldSeveralTimes as well as when replacing it with pure ().

I don't see that---all tests fail if I change that definition. Try again with my latest pushed commit?

Reply:
> > Then, the tests pass both with the current definition of yieldSeveralTimes as well as when replacing it with pure ().
> 
> I don't see that---all tests fail if I change that definition. Try again with my latest pushed commit?

Ah yeah, I commented out the `ExtendSelection` command for debugging and then forgot to add it back when testing with/without `yieldSeveralTimes` :facepalm: Now get the failures as expected :+1:

---

Minor: The regression test could use `withMaxSuccesses 1`.